### ISSUE TITLE
Possible bug in A2C/ACKTR?

### ISSUE THREAD

**Issue by siddk** [ikostrikov/pytorch-a2c-ppo-acktr#146]
I've been going through the update code for A2C in master/algo/a2c_acktr.py here:

```python
def update(self, rollouts):
        obs_shape = rollouts.obs.size()[2:]
        action_shape = rollouts.actions.size()[-1]
        num_steps, num_processes, _ = rollouts.rewards.size()
        values, action_log_probs, dist_entropy, _ = self.actor_critic.evaluate_actions(
            rollouts.obs[:-1].view(-1, *obs_shape),
            rollouts.recurrent_hidden_states[0].view(-1, self.actor_critic.recurrent_hidden_state_size),
            rollouts.masks[:-1].view(-1, 1),
            rollouts.actions.view(-1, action_shape))

        values = values.view(num_steps, num_processes, 1)
        action_log_probs = action_log_probs.view(num_steps, num_processes, 1)

        advantages = rollouts.returns[:-1] - values
        value_loss = advantages.pow(2).mean()

        action_loss = -(advantages.detach() * action_log_probs).mean()

        if self.acktr and self.optimizer.steps % self.optimizer.Ts == 0:
            ...

        self.optimizer.zero_grad()
        (value_loss * self.value_loss_coef + action_loss -
         dist_entropy * self.entropy_coef).backward()

        if self.acktr == False:
            nn.utils.clip_grad_norm_(self.actor_critic.parameters(),
                                     self.max_grad_norm)

        self.optimizer.step()

        return value_loss.item(), action_loss.item(), dist_entropy.item()
```

While the masks are being passed to self.actor_critic.evaluate_actions - it doesn't seem like they're being used at all in masking the action_loss. 

In other words, if I have 8 workers, and 6 of them terminate after 2 steps, and the remaining 2 workers continue rolling out for the full 20 steps, it seems as if I'm performing gradient updates for the actions taken predicted by the terminated workers (6 * (20-2)) additional steps)?

Seems like this could be resolved by doing the following:

```python
action_loss = -(advantages.detach() * action_log_probs * rollout.masks[:-1])
```

but I'm not sure. Is there another place in the code that this masking happens?

**Comment by ikostrikov** [comment]
VecNorm wrapper restarts workers automatically:
https://github.com/openai/baselines/blob/master/baselines/common/vec_env/subproc_vec_env.py#L13

Masks are used to compute returns:
https://github.com/ikostrikov/pytorch-a2c-ppo-acktr/blob/master/main.py#L119
https://github.com/ikostrikov/pytorch-a2c-ppo-acktr/blob/master/storage.py#L55

**Comment by siddk** [comment]
Got it - thanks!
